titus-common/src/main/java/com/netflix/titus/common/util/loadshedding/tokenbucket/TokenBucketAdmissionController.java

```diff
@@ -0,0 +1,146 @@
+ * limitations under the License.
+ */
+
+package com.netflix.titus.common.util.loadshedding.tokenbucket;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.cache.Cache;
+import com.netflix.titus.common.util.cache.Caches;
+import com.netflix.titus.common.util.limiter.Limiters;
+import com.netflix.titus.common.util.limiter.tokenbucket.TokenBucket;
+import com.netflix.titus.common.util.loadshedding.AdmissionController;
+import com.netflix.titus.common.util.loadshedding.AdmissionControllerRequest;
+import com.netflix.titus.common.util.loadshedding.AdmissionControllerResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Admission controller with multiple token buckets. A token bucket is selected by evaluating matching criteria in
+ * declaration order until first match is found. The selected bucket is tried, and either success or failure is
+ * returned to the caller.
+ */
+public class TokenBucketAdmissionController implements AdmissionController {
+
+    private static final Logger logger = LoggerFactory.getLogger(TokenBucketAdmissionController.class);
+
+    private static final AdmissionControllerResponse DEFAULT_OK = AdmissionControllerResponse.newBuilder()
+            .withAllowed(true)
+            .withReasonMessage("Rate limits not configured")
+            .build();
+
+    private static final int MAX_CACHE_SIZE = 10_000;
+    private static final Duration CACHE_ITEM_TIMEOUT = Duration.ofSeconds(60);
+
+    private final List<TokenBucketConfiguration> tokenBucketConfigurations;
+    private final Cache<String, TokenBucketInstance> bucketsById;
+
+    public TokenBucketAdmissionController(List<TokenBucketConfiguration> tokenBucketConfigurations,
+                                          TitusRuntime titusRuntime) {
+        this.tokenBucketConfigurations = tokenBucketConfigurations;
+
+        this.bucketsById = Caches.instrumentedCacheWithMaxSize(
+                MAX_CACHE_SIZE,
+                CACHE_ITEM_TIMEOUT,
+                "titus.tokenBucketAdmissionController.cache",
+                titusRuntime.getRegistry()
+        );
+    }
+
+    @Override
+    public AdmissionControllerResponse apply(AdmissionControllerRequest request) {
+        return findTokenBucket(request).map(this::consume).orElse(DEFAULT_OK);
+    }
+
+    private Optional<TokenBucketInstance> findTokenBucket(AdmissionControllerRequest request) {
+        return tokenBucketConfigurations.stream()
+                .filter(tokenBucketConfiguration -> matches(tokenBucketConfiguration, request))
+                .findFirst()
+                .map(tokenBucketConfiguration -> {
+                    // If shared, use pattern name as key, otherwise use caller id
+                    String id = tokenBucketConfiguration.isShared()
```

Comment:
What if the same callerId has different configurations for different endpoints, don't we need a different `TokenBucketInstance` (id) for each `(callerPatternString, endpointPatternString)` pair?

Reply:
Thinking more about it, we also need `tokenBucketConfiguration.getName()` as part of the id here

Reply:
We need use callerId / endpoint name as bucket id.
`tokenBucketConfiguration.getName()` is not important